python/lsst/daf/butler/script/queryDatasets.py

```diff
@@ -214,16 +213,32 @@ def getDatasets(self) -> Iterator[DatasetRef]:
             Dataset references matching the given query criteria.
         """
         datasetTypes = self._dataset_type_glob or ...
-        query_collections: Iterable[str] | EllipsisType = self._collections_wildcard or ...
+        query_collections: Iterable[str] = self._collections_wildcard or ["*"]
 
         # Currently need to use old interface to get all the matching
         # dataset types and loop over the dataset types executing a new
         # query each time.
-        dataset_types = self.butler.registry.queryDatasetTypes(datasetTypes)
+        dataset_types: set[str] = {d.name for d in self.butler.registry.queryDatasetTypes(datasetTypes)}
         with self.butler._query() as query:
-            query_collections = self.butler.registry.queryCollections(query_collections)
+            # If there is more than one dataset type being requested include
+            # the summary information so we can pre-filter. With only one
+            # let query.datasets work it out.
+            include_summary = True if len(dataset_types) > 1 else False
+            query_collections_info = self.butler.collections.x_query_info(
+                query_collections, include_summary=include_summary
+            )
+            query_collections = [c.name for c in query_collections_info]
+
+            # Only iterate over dataset types that are relevant for the query.
+            if include_summary:
```

Comment:
@TallJimbo This change makes a huge difference to the "all dataset types in collection X" variant for `butler query-dataset-types`. Easily noticeable in ci_hsc_gen3 comparing the original query system version with this implementation. It doesn't seem to take long for the collection summaries to be calculated for ci_hsc so I'm not sure if "n > 1" dataset types is a reasonable cutoff. The code is in a position to report a reason for no results now for dataset types at least.

Reply:
This logic is needed elsewhere so I feel like I'm going to have to have an API somewhere that does:

```python
def filter_dataset_types(dataset_types: Iterable[str | DatasetType], collection_info: Iterable[CollectionInfo]) -> set[str]:
    ...
```
Although I can't think where that would go since both `Butler` and `ButlerCollections` seems wrong. Or else we add `collections` to the new `Butler.query_dataset_types` API and filter that way -- because, now that I think about it, another approach here is to ask the collections first for all the possible answers, then convert the dataset type expressions to regexes and filter the collections dataset types directly and then pass the filtered non-regex iterables to the call that converts them to DatasetType (it's unfortunate that the DatasetType objects are dropped on the floor by CollectionInfo in this case though).

Reply:
if it's not much of performance hit, it sounds like it makes sense to pull down summaries even for just one dataset type, for consistency.  I'm surprised that the new logic has a big performance impact - it suggests my intuition for was was expensive in the old system was off.

I do think we'll want something _like_ a `collections` argument to `query_dataset_types`, but for now it sounds like it'd be best to just put this in a package-private free function somewhere until we nail down the rest of the DM-43157 interfaces.

Reply:
Done. I changed the scripts to use it and even added support for `butler query-dataset-types` command line.